src/optimization.py

```diff
@@ -48,40 +66,46 @@ def __init__(self, par ,**kwargs):
         self.station_pow_max = event["pow_max"]
         self.station_pow_min = event["pow_min"]
         self.power_rate = event['power_rate']
-        self.dcm_charging_flex_params = np.array([[ - self.power_rate * 0.0184 / 2], [ self.power_rate * 0.0184 / 2], [0], [0]])
-        #% DCM parameters for choice 1 -- charging with flexibility
-        self.dcm_charging_asap_params = np.array([[self.power_rate * 0.0184 / 2], [- self.power_rate * 0.0184 / 2], [0],[0.341 ]])
-        #% DCM parameters for choice 2 -- charging as soon as possible
-        self.dcm_leaving_params = np.array([[self.power_rate * 0.005 / 2], [self.power_rate * 0.005 / 2], [0], [-1 ]])
-
-
-        
-        #% DCM parameters for choice 3 -- leaving without charging
+        self.dcm_charging_flex_params = np.array(
+            [[- self.power_rate * 0.0184 / 2], [self.power_rate * 0.0184 / 2], [0], [0]])
+        # % DCM parameters for choice 1 -- charging with flexibility
+        self.dcm_charging_asap_params = np.array(
+            [[self.power_rate * 0.0184 / 2], [- self.power_rate * 0.0184 / 2], [0], [0.341]])
+        # % DCM parameters for choice 2 -- charging as soon as possible
+        self.dcm_leaving_params = np.array([[self.power_rate * 0.005 / 2], [self.power_rate * 0.005 / 2], [0], [-1]])
+
+        # % DCM parameters for choice 3 -- leaving without charging
         self.THETA = np.vstack((self.dcm_charging_flex_params.T, self.dcm_charging_asap_params.T,
-                     self.dcm_leaving_params.T))
+                                self.dcm_leaving_params.T))
         # problem specifications
-        self.N_flex = self.user_duration # charging duration that is not charged, hour
-        
+        self.N_flex = self.user_duration  # charging duration that is not charged, hour
+
         ### IS THIS CORRECT? WHATS SOC NEED REPRESENTS? 
         # self.N_asap = math.floor((self.user_SOC_need - self.user_SOC_init) *
         #                          self.user_batt_cap / self.station_pow_max / par.eff / par.Ts)
 
         ## HERE 12 IS SELF CODED 
         self.N_asap = math.ceil((self.e_need / self.station_pow_max / par.eff * int(1 / par.Ts)))
 
+        if len(par.TOU) < self.user_time + self.user_duration:  # if there is overnight chaarging 
```

Comment:
if there is overnight ch*a*rging

Reply:
> if there is overnight ch_a_rging

Based on this logic, we will have 96 * 2 = 192 points for a length of 2 days. It is problematic when:
1) self.user_time is a global index, which means that it could be 4, 5, 6 days + hours, then it does not make sense;
2) self.user_time is a relative index, however, self.user_duration + self.user_time is greater than two days. Then it doesn't work.

I roughly fixed this part in the notebook by modifying the concatenated length.